wFix critical agent recording issues: layer init, downtrack race, debu…

am-sokolov opened the pull request:
…g logging

# Fix critical agent recording issues: layer init, downtrack race, debug logging

This commit applies three critical fixes for reliable agent recording from publisher tracks using manual subscription and explicit video quality requests.

## Issue 1: Fix spatial layer initialization in Forwarder

**Problem**: `SetMaxSpatialLayer()` correctly updated the max spatial layer but failed to initialize target and current layers when they were still set to `InvalidLayerSpatial`. This caused video forwarding to fail silently.

**Impact**: Agents using manual subscription with explicit video quality requests (`SetVideoQuality`) received 0 video packets despite successful subscription. Audio worked fine (separate code path), only video was affected.

**Fix**: Initialize target and current spatial layers when max is set and they remain invalid.

**File**: `pkg/sfu/forwarder.go`

```go
func (f *Forwarder) SetMaxSpatialLayer(spatialLayer int32) (bool, buffer.VideoLayer) {
    f.vls.SetMaxSpatial(spatialLayer)
    newMax := f.vls.GetMax()

    // Initialize target if invalid
    if f.vls.GetTarget().Spatial == buffer.InvalidLayerSpatial {
        f.vls.SetTarget(newMax)
        // Initialize current if invalid
        if f.vls.GetCurrent().Spatial == buffer.InvalidLayerSpatial {
            f.vls.SetCurrent(newMax)
        }
    }

    return true, newMax
}
```

## Issue 2: Fix WrappedReceiver downtrack race condition

**Problem**: When downtracks were added before the underlying `TrackReceiver` was ready (still `nil`), they were silently lost. This race condition occurred when agents subscribed immediately after track publication.

**Impact**: Recording agents and other clients subscribing quickly after track publish received no packets. The race window was small but reproducible in automated testing scenarios.

**Fix**: Queue downtracks in a pending list until receiver is ready, then flush the queue atomically.

**File**: `pkg/rtc/wrappedreceiver.go`

```go
type WrappedReceiver struct {
    // ... existing fields
    pendingDownTracks []sfu.TrackSender  // Queue for early subscribers
}

func (r *WrappedReceiver) AddDownTrack(track sfu.TrackSender) error {
    r.lock.Lock()
    defer r.lock.Unlock()

    if r.TrackReceiver != nil {
        return r.TrackReceiver.AddDownTrack(track)
    }

    // Queue until receiver ready
    r.pendingDownTracks = append(r.pendingDownTracks, track)
    return nil
}

// Flush queue when receiver becomes ready
func (r *WrappedReceiver) OnSetTrackReceiver() {
    for _, dt := range r.pendingDownTracks {
        r.TrackReceiver.AddDownTrack(dt)
    }
    r.pendingDownTracks = nil
}
```

## Issue 3: Add debug logging to RTP forwarding pipeline

**Problem**: Zero visibility into RTP packet forwarding made it nearly impossible to diagnose why clients weren't receiving packets. Issues #1 and #2 took hours to identify without logging.

**Impact**: Silent failures in video forwarding had no diagnostic information. Developers had to modify server code and rebuild to debug packet flow issues.

**Fix**: Add debug logging for the first 10 H.264 video packets showing:
- Packet details (sequence number, timestamp, payload size)
- Downtrack count before broadcast
- Write count (how many downtracks successfully received the packet)

**File**: `pkg/sfu/receiver.go`

## Combined Impact

These three issues created a "perfect storm" failure mode where:
- ✅ Client subscriptions succeeded (no errors thrown)
- ❌ Video packets never arrived (silent failure)
- ❌ No diagnostic information available (invisible problem)

**Affected Scenarios**:
- Recording agents using manual subscription + explicit video quality
- Any client calling `SetVideoQuality()` immediately after subscription
- Automated testing scenarios with tight timing
- Server-side recording of specific publisher tracks

**Unaffected Scenarios**:
- Standard participant-to-participant calls with auto-subscribe
- Clients using default video quality (no explicit layer setting)
- Subscriptions with delayed quality requests

## Why These Bugs Weren't Caught Earlier

The Python agents SDK uses `auto_subscribe=true` by default, which triggers a different subscription code path that avoids all three issues. These bugs specifically affect:
- Manual subscription workflows (`auto_subscribe=false`)
- Explicit video quality requests (`SetVideoQuality`)
- Immediate subscriptions after track publication

## Testing Validation

✅ Recording agents receive video packets immediately after subscription
✅ Multi-hour recordings complete without packet loss
✅ Immediate subscription after track publish works reliably
✅ Debug logs provide clear visibility into packet flow
✅ No regressions in standard participant subscriptions (verified)
✅ Works with both H.264 and VP8 video codecs

boks1971 commented:
Thank you for all the details @am-sokolov . Really appreciate it. Let me review this once again and get back to you. I think the defensive initialisation you have is probably fine. But, I wanted to look up historical notes to make sure that it does not introduce some unexpected behaviour for some cases.

am-sokolov commented:
Thank you for the feedback! I've updated the PR with all three changes:

  1. Added isDeficientLocked() check - Won't set target when congestion controller is throttling
  2. Removed SetCurrent() call - Current layer initialisation now deferred to keyframe-based forwarder start
  3. Changed to Debugw() logging - Single debug-level log to minimise noise

boks1971 commented:
Thank you very much @am-sokolov . Will merge this shortly.

boks1971 commented:
@am-sokolov Can you please sign the CLA so that we can accept this PR and merge? - https://cla-assistant.io/livekit/livekit?pullRequest=4003

boks1971 commented:
Thank you @am-sokolov . Really appreciate it.

I will take a look.
